### PR TITLE
(feat) Billable Exemption config

### PR DESCRIPTION
**Pull request by Ogollah**
## Requirements

- [ ] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [ ] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

### What does this PR do?
* This PR Billable Schema Exemption and preview feature
* View the existing billable schema exemption in a Tree View
* Edit and update the schema through a schema editor
* Allows for input of a dummy schema when no schema exists to ensure a user can quickly start editing and preview the exemptions in the preview tab.

## Screenshots

* Preview

Screenshot from 2024-12-02 22-08-04

*Schema editor
Screenshot from 2024-12-01 18-58-52

*None.*

## Related Issue

*None.*

## Other

*None.*

### CODE REVIEW

**Comment by ojwanganto** [comment]
Please share a screenshot of the new look.

**Comment by Ogollah** [comment]
The shared Preview screen is the new screen. I tried to maintain the same look as the previous one.

**Comment by ojwanganto** [comment]
> The shared Preview screen is the new screen. I tried to maintain the same look as the previous one. The edit schema screen did not change.

This is appropriate to the directory and files' presentation. You can use the default look from carbon.

**Comment by Ogollah** [comment]
Do you mean without the icons?

File: packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.component.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        />
+      </div>
+    </div>
+  );
+
+  return (
+    <div>
+      {errors.length && validationOn ? <ErrorMessages /> : null}
+      <AceEditor
+        style={{ height: '100vh', width: '100%', border: errors.length ? '3px solid #DA1E28' : 'none' }}
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.component.tsx:152]: We should extract these syles to `scss` file

File: packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.scss [hunk 2]
```diff
@@ -0,0 +1,31 @@
+@use '@carbon/colors';
+@use '@carbon/type';
+
+.validationErrorsContainer {
+  display: flex;
+  gap: 0.5rem;
+
+  :global(.cds--actionable-notification) {
+    max-width: 38rem;
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.scss:9]: Will this be responsive on a small screen

File: packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.scss [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  }
+  :global(.cds--actionable-notification__details) {
+    margin: 0 0rem 0 0.8125rem !important;
+  }
+}
+
+.pagination {
+  display: flex;
+  align-items: center;
+  gap: 0.5rem;
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-editor.scss:19]: Lets use the carbon tokens for size

File: packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-viewer-component.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+interface TreeNode {
+  id: string;
+  label: string;
+  children?: TreeNode[];
+  icon: React.ReactNode;
+}
+
+const transformDataToTree = (data: Record<string, any>, parentKey = 'root'): TreeNode[] => {
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/billable-services/billable-exemptions/schema-editor/schema-viewer-component.tsx:17]:
```suggestion
const transformDataToTree = (data: Record<string, any>, parentKey = 'root'): Array<TreeNode> => {
```

File: packages/esm-billing-app/src/hooks/useStandardSchema.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import useSWRImmutable from 'swr/immutable';
+import { openmrsFetch } from '@openmrs/esm-framework';
+
+export function useStandardSchema(key: string) {
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/hooks/useStandardSchema.ts:4]: should we rename this to `useExemptionSchema`

File: packages/esm-billing-app/src/hooks/useSystemBillableSetting.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+export async function saveExemptionSchema(schema: Schema, key: string): Promise<FetchResponse<Schema>> {
+  const body: SavePayload = { value: JSON.stringify(schema) };
+  const headers = { 'Content-Type': 'application/json' };
+
+  return openmrsFetch(`/ws/rest/v1/systemsetting/${key}`, {
+    method: 'POST',
+    headers,
+    body,
+  });
+}
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/hooks/useSystemBillableSetting.tsx:35]: should we combine the functionality in `useStandardSchema.tsx` and this file to one hook. They all access and manipulate the same endpoint

File: packages/esm-billing-app/src/root.component.tsx [hunk 7]
```diff
@@ -46,6 +47,7 @@ const RootComponent: React.FC = () => {
         <Route path="/bill-manager" element={<BillManager />} />
         <Route path="/charge-items" element={<ChargeItemsDashboard />} />
         <Route path="/payment-modes" element={<PaymentModeHome />} />
+        <Route path="/billable-exemptions" element={<BillableExemptions />} />
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/root.component.tsx:50]: After this PR is merged. Add a ticket to add control access by adding correct config for it

File: packages/esm-billing-app/src/types/index.ts [hunk 8]
```diff
@@ -548,3 +548,19 @@ export type FacilityClaim = {
   externalId: string;
 };
 export type BillingPromptType = 'patient-chart' | 'billing-orders';
+
+export interface Schema {
+  services: {
+    all: Service[];
```
Comment (donaldkibet) [inline packages/esm-billing-app/src/types/index.ts:554]:
```suggestion
    all: Array<Service>
```
its a nit but that the convention we are using on the project

**Review by donaldkibet — APPROVED** [review]

**Comment by donaldkibet** [comment]
@Ogollah merging this in to  test it out. you can follow up with PR to fix the recommended changes

**Review by makombe — APPROVED** [review]

**Pull request merged**